In [8]:
import os
from pathlib import Path

dataset_path = Path(".")

folders = sorted([f for f in dataset_path.iterdir() if f.is_dir() and f.name != "venv"])
for folder in folders:
    json_files = list(folder.rglob("*.json"))
    print(f"{folder.name}  —  {len(json_files)} file(s)")

aws_dataset  —  55 file(s)
botsv3  —  0 file(s)


In [9]:
import json

# Grab the first JSON file from the first folder
first_file = list(folders[0].rglob("*.json"))[0]
print(f"Opening: {first_file}\n")

with open(first_file) as f:
    data = json.load(f)

records = data.get("Records", [])
print(f"Number of events in this file: {len(records)}")
print()
print(json.dumps(records[0], indent=2))

Opening: aws_dataset\CloudTrail\218007301253_CloudTrail_us-east-1_20230710T1145Z_7xgocspSowgK0Gto.json

Number of events in this file: 29

{
  "eventVersion": "1.08",
  "userIdentity": {
    "type": "IAMUser",
    "principalId": "AIDATFQR7NSC5U6Q3TMDR",
    "arn": "arn:aws:iam::123837392027:user/benjamin",
    "accountId": "123837392027",
    "accessKeyId": "ASIATFQR7NSCTM3XVF6B",
    "userName": "benjamin",
    "sessionContext": {
      "sessionIssuer": {},
      "webIdFederationData": {},
      "attributes": {
        "creationDate": "2023-07-10T11:42:31Z",
        "mfaAuthenticated": "true"
      }
    },
    "invokedBy": "AWS Internal"
  },
  "eventTime": "2023-07-10T11:42:36Z",
  "eventSource": "s3.amazonaws.com",
  "eventName": "GetStorageLensConfiguration",
  "awsRegion": "us-east-1",
  "sourceIPAddress": "AWS Internal",
  "userAgent": "AWS Internal",
  "requestParameters": {
    "Host": "123837392027.s3-control.us-east-1.amazonaws.com"
  },
  "responseElements": null,
  "additi

In [10]:
import json
import pandas as pd
from pathlib import Path
from collections import Counter

cloudtrail_path = Path("./aws_dataset/CloudTrail")

all_events = []
event_names = Counter()
identity_types = Counter()

for json_file in sorted(cloudtrail_path.rglob("*.json")):
    with open(json_file) as f:
        data = json.load(f)
    
    records = data.get("Records", [])
    for event in records:
        identity = event.get("userIdentity", {})
        row = {
            "timestamp":      event.get("eventTime"),
            "event_name":     event.get("eventName"),
            "event_source":   event.get("eventSource"),
            "principal_type": identity.get("type"),
            "principal_arn":  identity.get("arn"),
            "username":       identity.get("userName"),
            "source_ip":      event.get("sourceIPAddress"),
            "error_code":     event.get("errorCode"),
            "aws_region":     event.get("awsRegion"),
            "source_file":    json_file.name,
        }
        all_events.append(row)
        event_names[event.get("eventName", "unknown")] += 1
        identity_types[identity.get("type", "missing")] += 1

df = pd.DataFrame(all_events)
df["timestamp"] = pd.to_datetime(df["timestamp"])
df.sort_values("timestamp", inplace=True)
df.reset_index(drop=True, inplace=True)

print(f"Total events loaded: {len(df)}")
print(f"\nDate range: {df['timestamp'].min()}  →  {df['timestamp'].max()}")
print(f"\nidentity types:\n{identity_types}")
print(f"\nTop 150 API calls:")
for name, count in event_names.most_common(150):
    print(f"  {name}: {count}")

Total events loaded: 2900

Date range: 2023-07-10 11:42:18+00:00  →  2023-07-10 12:37:50+00:00

identity types:
Counter({'IAMUser': 2748, 'AssumedRole': 76, 'missing': 42, 'AWSService': 34})

Top 150 API calls:
  Decrypt: 178
  DescribeRouteTables: 163
  GetUser: 130
  DescribeParameters: 122
  ListTagsForResource: 88
  GetParameter: 82
  DeleteParameter: 78
  PutParameter: 67
  GetSecretValue: 60
  DescribeNatGateways: 54
  AssumeRole: 49
  DescribeEventAggregates: 48
  DescribeVpcAttribute: 48
  DescribeOrderableDBInstanceOptions: 45
  DescribeVpcs: 43
  GetBucketAcl: 42
  Encrypt: 42
  DescribeAccountAttributes: 41
  DescribeSubnets: 40
  ListAttachedRolePolicies: 39
  GetResourcePolicy: 39
  DescribeSecret: 36
  DescribeDBInstances: 35
  ListRolePolicies: 33
  GetRole: 31
  DescribeSecurityGroups: 30
  GetPasswordData: 29
  DescribeInstanceInformation: 27
  DescribeAvailabilityZones: 26
  DescribeNetworkAcls: 26
  DescribeInstanceAttribute: 25
  DescribeInternetGateways: 24
  Descr

In [11]:
iam_events = df[df["event_source"] == "iam.amazonaws.com"]
print(f"IAM events: {len(iam_events)}")
print()
print(iam_events["event_name"].value_counts().to_string())

IAM events: 398

event_name
GetUser                           130
ListAttachedRolePolicies           39
ListRolePolicies                   33
GetRole                            31
CreateRole                         13
ListInstanceProfilesForRole        13
DeleteRole                         13
GetRolePolicy                      11
GetPolicy                           8
GetInstanceProfile                  7
AttachRolePolicy                    6
PutRolePolicy                       5
DetachRolePolicy                    5
ListAccessKeys                      5
DeleteLoginProfile                  5
ListSSHPublicKeys                   4
ListMFADevices                      4
GetPolicyVersion                    4
ListEntitiesForPolicy               4
DeleteRolePolicy                    4
CreateUser                          4
DeleteUser                          4
CreateInstanceProfile               3
AddRoleToInstanceProfile            3
TagInstanceProfile                  3
RemoveRoleFromInstance

In [12]:
import json
import pandas as pd
from pathlib import Path

cloudtrail_path = Path("./aws_dataset/CloudTrail")

# Known attack-associated IAM/privilege escalation API calls
# Sourced from Stratus Red Team technique documentation
ATTACK_EVENTS = {
    # Privilege escalation
    "CreateAccessKey":          "privilege-escalation",
    "CreateLoginProfile":       "privilege-escalation",
    "UpdateLoginProfile":       "privilege-escalation",
    "AttachUserPolicy":         "privilege-escalation",
    "AttachRolePolicy":         "privilege-escalation",
    "AttachGroupPolicy":        "privilege-escalation",
    "PutUserPolicy":            "privilege-escalation",
    "PutRolePolicy":            "privilege-escalation",
    "PutGroupPolicy":           "privilege-escalation",
    "CreatePolicyVersion":      "privilege-escalation",
    "SetDefaultPolicyVersion":  "privilege-escalation",
    "AddUserToGroup":           "privilege-escalation",
    
    # Persistence
    "CreateUser":               "persistence",
    "CreateRole":               "persistence",
    "UpdateAssumeRolePolicy":   "persistence",
    
    # Defense evasion
    "StopLogging":              "defense-evasion",
    "DeleteTrail":              "defense-evasion",
    "UpdateTrail":              "defense-evasion",
    "PutEventSelectors":        "defense-evasion",
    
    # Credential access
    "GetSecretValue":           "credential-access",
    "GetPasswordData":          "credential-access",
    
    # Exfiltration
    "PutBucketPolicy":          "exfiltration",
    "DeleteBucketPolicy":       "exfiltration",
}

all_events = []

for json_file in sorted(cloudtrail_path.rglob("*.json")):
    with open(json_file) as f:
        data = json.load(f)
    
    for event in data.get("Records", []):
        identity = event.get("userIdentity", {})
        event_name = event.get("eventName", "unknown")
        
        row = {
            "timestamp":        event.get("eventTime"),
            "event_name":       event_name,
            "event_source":     event.get("eventSource"),
            "principal_type":   identity.get("type"),
            "principal_arn":    identity.get("arn"),
            "username":         identity.get("userName"),
            "source_ip":        event.get("sourceIPAddress"),
            "error_code":       event.get("errorCode"),
            "aws_region":       event.get("awsRegion"),
            "label":            1 if event_name in ATTACK_EVENTS else 0,
            "attack_technique": ATTACK_EVENTS.get(event_name, None),
        }
        all_events.append(row)

df = pd.DataFrame(all_events)
df["timestamp"] = pd.to_datetime(df["timestamp"])
df.sort_values("timestamp", inplace=True)
df.reset_index(drop=True, inplace=True)

print(f"Total events:  {len(df)}")
print(f"Benign (0):    {(df['label'] == 0).sum()}")
print(f"Attack (1):    {(df['label'] == 1).sum()}")
print(f"\nAttack technique breakdown:")
print(df[df['label']==1]['attack_technique'].value_counts())
print(f"\nAll unique event names seen:")
print(df['event_name'].value_counts().to_string())

Total events:  2900
Benign (0):    2764
Attack (1):    136

Attack technique breakdown:
attack_technique
credential-access       89
persistence             19
privilege-escalation    16
defense-evasion          8
exfiltration             4
Name: count, dtype: int64

All unique event names seen:
event_name
Decrypt                                     178
DescribeRouteTables                         163
GetUser                                     130
DescribeParameters                          122
ListTagsForResource                          88
GetParameter                                 82
DeleteParameter                              78
PutParameter                                 67
GetSecretValue                               60
DescribeNatGateways                          54
AssumeRole                                   49
DescribeEventAggregates                      48
DescribeVpcAttribute                         48
DescribeOrderableDBInstanceOptions           45
DescribeVpcs         

In [13]:
print(df.isnull().sum())

timestamp              0
event_name             0
event_source           0
principal_type        42
principal_arn         77
username             152
source_ip              0
error_code          2600
aws_region             0
label                  0
attack_technique    2764
dtype: int64


In [14]:
print(df['principal_type'].value_counts())

principal_type
IAMUser        2748
AssumedRole      76
AWSService       34
Name: count, dtype: int64


In [15]:
attack_df = df[df['label'] == 1][['timestamp', 'username', 'event_name', 'attack_technique', 'error_code']]
print(attack_df.to_string())

                     timestamp  username              event_name      attack_technique                    error_code
87   2023-07-10 11:54:39+00:00  bert-jan              CreateRole           persistence                           NaN
89   2023-07-10 11:54:39+00:00  bert-jan           PutRolePolicy  privilege-escalation                           NaN
98   2023-07-10 11:54:47+00:00       NaN         GetPasswordData     credential-access  Client.UnauthorizedOperation
99   2023-07-10 11:54:47+00:00       NaN         GetPasswordData     credential-access  Client.UnauthorizedOperation
100  2023-07-10 11:54:47+00:00       NaN         GetPasswordData     credential-access  Client.UnauthorizedOperation
102  2023-07-10 11:54:47+00:00       NaN         GetPasswordData     credential-access  Client.UnauthorizedOperation
103  2023-07-10 11:54:48+00:00       NaN         GetPasswordData     credential-access  Client.UnauthorizedOperation
104  2023-07-10 11:54:48+00:00       NaN         GetPasswordData

In [16]:
df.to_csv("invictus_labeled.csv", index=False)
print("Saved. Shape:", df.shape)

Saved. Shape: (2900, 11)


In [17]:
def normalise_principal(identity: dict) -> dict:
    """Extract clean principal info regardless of userIdentity type."""
    id_type = identity.get("type", "unknown")
    
    if id_type == "IAMUser":
        return {
            "principal_type": "IAMUser",
            "principal_arn":  identity.get("arn"),
            "username":       identity.get("userName"),
        }
    
    elif id_type == "AssumedRole":
        # ARN is nested inside sessionContext for assumed roles
        session = identity.get("sessionContext", {})
        issuer = session.get("sessionIssuer", {})
        return {
            "principal_type": "AssumedRole",
            "principal_arn":  identity.get("arn"),          # the assumed session ARN
            "username":       issuer.get("userName"),         # the original role name
        }
    
    elif id_type == "AWSService":
        return {
            "principal_type": "AWSService",
            "principal_arn":  None,
            "username":       identity.get("invokedBy"),     # e.g. "ec2.amazonaws.com"
        }
    
    else:
        return {
            "principal_type": id_type,
            "principal_arn":  identity.get("arn"),
            "username":       identity.get("userName"),
        }


# Test it on a few raw events to verify
import json
from pathlib import Path

test_file = list(Path("./aws_dataset/CloudTrail").rglob("*.json"))[0]
with open(test_file) as f:
    data = json.load(f)

for event in data["Records"][:5]:
    identity = event.get("userIdentity", {})
    result = normalise_principal(identity)
    print(f"{event['eventName']:30} → {result}")

GetStorageLensConfiguration    → {'principal_type': 'IAMUser', 'principal_arn': 'arn:aws:iam::123837392027:user/benjamin', 'username': 'benjamin'}
GetBucketPublicAccessBlock     → {'principal_type': 'IAMUser', 'principal_arn': 'arn:aws:iam::123837392027:user/benjamin', 'username': 'benjamin'}
GetBucketPolicyStatus          → {'principal_type': 'IAMUser', 'principal_arn': 'arn:aws:iam::123837392027:user/benjamin', 'username': 'benjamin'}
GetBucketAcl                   → {'principal_type': 'IAMUser', 'principal_arn': 'arn:aws:iam::123837392027:user/benjamin', 'username': 'benjamin'}
GetBucketPublicAccessBlock     → {'principal_type': 'IAMUser', 'principal_arn': 'arn:aws:iam::123837392027:user/benjamin', 'username': 'benjamin'}


In [18]:
all_events = []

for json_file in sorted(Path("./aws_dataset/CloudTrail").rglob("*.json")):
    with open(json_file) as f:
        data = json.load(f)
    
    for event in data.get("Records", []):
        identity = event.get("userIdentity", {})
        principal = normalise_principal(identity)
        
        event_name = event.get("eventName", "unknown")
        
        row = {
            "timestamp":        event.get("eventTime"),
            "event_name":       event_name,
            "event_source":     event.get("eventSource"),
            "aws_region":       event.get("awsRegion"),
            "source_ip":        event.get("sourceIPAddress"),
            "error_code":       event.get("errorCode"),
            "label":            1 if event_name in ATTACK_EVENTS else 0,
            "attack_technique": ATTACK_EVENTS.get(event_name),
            **principal,        # unpacks principal_type, principal_arn, username
        }
        all_events.append(row)

df_clean = pd.DataFrame(all_events)
df_clean["timestamp"] = pd.to_datetime(df_clean["timestamp"])
df_clean.sort_values("timestamp", inplace=True)
df_clean.reset_index(drop=True, inplace=True)

# Verify the schema
print(df_clean.dtypes)
print(f"\nShape: {df_clean.shape}")
print(f"\nNull counts:\n{df_clean.isnull().sum()}")
print(f"\nLabel split:\n{df_clean['label'].value_counts()}")

timestamp           datetime64[us, UTC]
event_name                          str
event_source                        str
aws_region                          str
source_ip                           str
error_code                          str
label                             int64
attack_technique                    str
principal_type                      str
principal_arn                       str
username                            str
dtype: object

Shape: (2900, 11)

Null counts:
timestamp              0
event_name             0
event_source           0
aws_region             0
source_ip              0
error_code          2600
label                  0
attack_technique    2764
principal_type         0
principal_arn         77
username              42
dtype: int64

Label split:
label
0    2764
1     136
Name: count, dtype: int64


In [19]:
df_clean.to_csv("invictus_labeled_clean.csv", index=False)
print("Saved invictus_labeled_clean.csv")
print(df_clean.head(3).to_string())

Saved invictus_labeled_clean.csv
                  timestamp          event_name           event_source aws_region     source_ip error_code  label attack_technique principal_type                            principal_arn  username
0 2023-07-10 11:42:18+00:00  GetRegionOptStatus  account.amazonaws.com  us-east-1  10.248.16.43        NaN      0              NaN        IAMUser  arn:aws:iam::123837392027:user/benjamin  benjamin
1 2023-07-10 11:42:23+00:00    GetBucketLogging       s3.amazonaws.com  us-east-1  10.248.16.43        NaN      0              NaN        IAMUser  arn:aws:iam::123837392027:user/benjamin  benjamin
2 2023-07-10 11:42:23+00:00     GetBucketPolicy       s3.amazonaws.com  us-east-1  10.248.16.43        NaN      0              NaN        IAMUser  arn:aws:iam::123837392027:user/benjamin  benjamin


In [20]:
def normalise_principal(identity: dict) -> dict:
    id_type = identity.get("type", "unknown")
    
    if id_type == "IAMUser":
        return {
            "principal_type": "IAMUser",
            "principal_arn":  identity.get("arn"),
            "username":       identity.get("userName"),
        }
    elif id_type == "AssumedRole":
        session = identity.get("sessionContext", {})
        issuer  = session.get("sessionIssuer", {})
        return {
            "principal_type": "AssumedRole",
            "principal_arn":  identity.get("arn"),
            "username":       issuer.get("userName"),
        }
    elif id_type == "AWSService":
        return {
            "principal_type": "AWSService",
            "principal_arn":  None,
            "username":       identity.get("invokedBy"),
        }
    else:
        return {
            "principal_type": id_type,
            "principal_arn":  identity.get("arn"),
            "username":       identity.get("userName"),
        }


# Rebuild with proper normalisation
all_events = []

for json_file in sorted(cloudtrail_path.rglob("*.json")):
    with open(json_file) as f:
        data = json.load(f)
    for event in data.get("Records", []):
        identity   = event.get("userIdentity", {})
        principal  = normalise_principal(identity)
        event_name = event.get("eventName", "unknown")
        row = {
            "timestamp":        event.get("eventTime"),
            "event_name":       event_name,
            "event_source":     event.get("eventSource"),
            "aws_region":       event.get("awsRegion"),
            "source_ip":        event.get("sourceIPAddress"),
            "error_code":       event.get("errorCode"),
            "label":            1 if event_name in ATTACK_EVENTS else 0,
            "attack_technique": ATTACK_EVENTS.get(event_name),
            **principal,
        }
        all_events.append(row)

df = pd.DataFrame(all_events)
df["timestamp"] = pd.to_datetime(df["timestamp"])
df.sort_values("timestamp", inplace=True)
df.reset_index(drop=True, inplace=True)

# Session labeling — propagate attack label ±5 minutes around each attack event
WINDOW_MINUTES = 5
df["session_label"] = df["label"].copy()

for username, group in df.groupby("username"):
    attack_times = group[group["label"] == 1]["timestamp"]
    for atk_time in attack_times:
        mask = (
            (df["username"] == username) &
            (df["timestamp"] >= atk_time - pd.Timedelta(minutes=WINDOW_MINUTES)) &
            (df["timestamp"] <= atk_time + pd.Timedelta(minutes=WINDOW_MINUTES))
        )
        df.loc[mask, "session_label"] = 1

print("Event-level attacks:    ", df["label"].sum())
print("Session-level attacks:  ", df["session_label"].sum())
print("\nSession label by username:")
print(df.groupby("username")[["label", "session_label"]].sum())
print("\nColumns:", list(df.columns))

Event-level attacks:     136
Session-level attacks:   2597

Session label by username:
                                             label  session_label
username                                                         
AWSServiceRoleForAmazonInspector2                0              0
AWSServiceRoleForRDS                             0              0
benjamin                                         0              0
bert-jan                                       107           2568
cloudtrail.amazonaws.com                         0              0
ec2.amazonaws.com                                0              0
inspector2.amazonaws.com                         0              0
lambda.amazonaws.com                             0              0
rds.amazonaws.com                                0              0
rolesanywhere.amazonaws.com                      0              0
stratus-red-team-ec2-enumerate-role              0              0
stratus-red-team-ec2-get-password-data-role     29     

In [21]:
# Save final dataset
df.to_csv("invictus_labeled_final.csv", index=False)
print("Saved invictus_labeled_final.csv")
print(f"Shape: {df.shape}")
print(f"Columns: {list(df.columns)}")

# Summary stats for the team
print("\n" + "="*50)
print("DATASET SUMMARY — invictus-ir")
print("="*50)
print(f"\nTotal events:          {len(df)}")
print(f"Unique users:          {df['username'].nunique()}")
print(f"Date range:            {df['timestamp'].min()} → {df['timestamp'].max()}")
print(f"Duration:              {df['timestamp'].max() - df['timestamp'].min()}")

print(f"\nLabel distribution (event-level):")
print(f"  Benign  (0): {(df['label']==0).sum()} ({(df['label']==0).mean()*100:.1f}%)")
print(f"  Attack  (1): {(df['label']==1).sum()} ({(df['label']==1).mean()*100:.1f}%)")

print(f"\nLabel distribution (session-level):")
print(f"  Benign  (0): {(df['session_label']==0).sum()} ({(df['session_label']==0).mean()*100:.1f}%)")
print(f"  Attack  (1): {(df['session_label']==1).sum()} ({(df['session_label']==1).mean()*100:.1f}%)")

print(f"\nAttack technique breakdown (event-level):")
print(df[df['label']==1]['attack_technique'].value_counts().to_string())

print(f"\nPrincipal type breakdown:")
print(df['principal_type'].value_counts().to_string())

print(f"\nTop 10 event sources:")
print(df['event_source'].value_counts().head(10).to_string())

print(f"\nNull counts:")
print(df.isnull().sum().to_string())

Saved invictus_labeled_final.csv
Shape: (2900, 12)
Columns: ['timestamp', 'event_name', 'event_source', 'aws_region', 'source_ip', 'error_code', 'label', 'attack_technique', 'principal_type', 'principal_arn', 'username', 'session_label']

DATASET SUMMARY — invictus-ir

Total events:          2900
Unique users:          18
Date range:            2023-07-10 11:42:18+00:00 → 2023-07-10 12:37:50+00:00
Duration:              0 days 00:55:32

Label distribution (event-level):
  Benign  (0): 2764 (95.3%)
  Attack  (1): 136 (4.7%)

Label distribution (session-level):
  Benign  (0): 303 (10.4%)
  Attack  (1): 2597 (89.6%)

Attack technique breakdown (event-level):
attack_technique
credential-access       89
persistence             19
privilege-escalation    16
defense-evasion          8
exfiltration             4

Principal type breakdown:
principal_type
IAMUser        2748
AssumedRole      76
unknown          42
AWSService       34

Top 10 event sources:
event_source
ec2.amazonaws.com         

In [22]:
# Find the unknown principal types
import json
from pathlib import Path

unknown_identities = []
for json_file in sorted(cloudtrail_path.rglob("*.json")):
    with open(json_file) as f:
        data = json.load(f)
    for event in data.get("Records", []):
        identity = event.get("userIdentity", {})
        if identity.get("type") not in ("IAMUser", "AssumedRole", "AWSService", None):
            unknown_identities.append({
                "type":       identity.get("type"),
                "event_name": event.get("eventName"),
                "arn":        identity.get("arn"),
                "raw":        identity
            })

print(f"Unknown identity count: {len(unknown_identities)}")
for u in unknown_identities[:5]:
    print(f"\ntype: {u['type']}")
    print(f"event: {u['event_name']}")
    import json as j
    print(j.dumps(u['raw'], indent=2))

Unknown identity count: 0


## Enriched Extraction

Re-extracts all CloudTrail events with additional fields needed for graph construction and sequence modeling:

| New column | Source | Why it matters |
|---|---|---|
| `read_only` | `readOnly` | All attack write ops are `False`; quick filter feature |
| `user_agent` | `userAgent` | Stratus Red Team leaves a fingerprint; also distinguishes Terraform from manual |
| `access_key_id` | `userIdentity.accessKeyId` | Tracks which credential was used; links sessions across events |
| `mfa_authenticated` | `userIdentity.sessionContext.attributes.mfaAuthenticated` | Security posture signal |
| `target_resource` | parsed from `requestParameters` | The *target* of the API call — essential for graph edges (who acted on what) |
| `request_params_raw` | `requestParameters` (JSON string) | Full parameters for downstream parsing |

In [23]:
import json
import pandas as pd
from pathlib import Path

cloudtrail_path = Path("./aws_dataset/CloudTrail")

# ── Attack label map (unchanged from previous cells) ──────────────────────────
ATTACK_EVENTS = {
    "CreateAccessKey":         "privilege-escalation",
    "CreateLoginProfile":      "privilege-escalation",
    "UpdateLoginProfile":      "privilege-escalation",
    "AttachUserPolicy":        "privilege-escalation",
    "AttachRolePolicy":        "privilege-escalation",
    "AttachGroupPolicy":       "privilege-escalation",
    "PutUserPolicy":           "privilege-escalation",
    "PutRolePolicy":           "privilege-escalation",
    "PutGroupPolicy":          "privilege-escalation",
    "CreatePolicyVersion":     "privilege-escalation",
    "SetDefaultPolicyVersion": "privilege-escalation",
    "AddUserToGroup":          "privilege-escalation",
    "CreateUser":              "persistence",
    "CreateRole":              "persistence",
    "UpdateAssumeRolePolicy":  "persistence",
    "StopLogging":             "defense-evasion",
    "DeleteTrail":             "defense-evasion",
    "UpdateTrail":             "defense-evasion",
    "PutEventSelectors":       "defense-evasion",
    "GetSecretValue":          "credential-access",
    "GetPasswordData":         "credential-access",
    "PutBucketPolicy":         "exfiltration",
    "DeleteBucketPolicy":      "exfiltration",
}

# ── Helpers ────────────────────────────────────────────────────────────────────

def normalise_principal(identity: dict) -> dict:
    id_type = identity.get("type", "unknown")
    if id_type == "IAMUser":
        return {
            "principal_type": "IAMUser",
            "principal_arn":  identity.get("arn"),
            "username":       identity.get("userName"),
        }
    elif id_type == "AssumedRole":
        session = identity.get("sessionContext", {})
        issuer  = session.get("sessionIssuer", {})
        return {
            "principal_type": "AssumedRole",
            "principal_arn":  identity.get("arn"),
            "username":       issuer.get("userName"),
        }
    elif id_type == "AWSService":
        return {
            "principal_type": "AWSService",
            "principal_arn":  None,
            "username":       identity.get("invokedBy"),
        }
    else:
        return {
            "principal_type": id_type or "unknown",
            "principal_arn":  identity.get("arn"),
            "username":       identity.get("userName"),
        }


# Keys to try, in priority order, when extracting the target resource from
# requestParameters.  The first key that exists and has a non-empty value wins.
_TARGET_KEYS = [
    "roleName",        # IAM role operations
    "userName",        # IAM user operations
    "groupName",       # IAM group operations
    "policyArn",       # Attach/detach policy operations
    "bucketName",      # S3 bucket operations  (also appears as "Host" for path-style)
    "secretId",        # Secrets Manager
    "instanceId",      # EC2
    "functionName",    # Lambda
    "trailName",       # CloudTrail
    "keyId",           # KMS
    "dbInstanceIdentifier",  # RDS
]

def extract_target_resource(params) -> str | None:
    """Return the primary resource name/ARN from requestParameters."""
    if not params or not isinstance(params, dict):
        return None
    for key in _TARGET_KEYS:
        val = params.get(key)
        if val:
            return str(val)
    # Fallback: return the first non-empty string value in the dict
    for val in params.values():
        if isinstance(val, str) and val:
            return val
    return None


# ── Main extraction loop ───────────────────────────────────────────────────────

all_events = []

for json_file in sorted(cloudtrail_path.rglob("*.json")):
    with open(json_file) as f:
        data = json.load(f)

    for event in data.get("Records", []):
        identity   = event.get("userIdentity", {})
        principal  = normalise_principal(identity)
        event_name = event.get("eventName", "unknown")
        params     = event.get("requestParameters")
        session    = identity.get("sessionContext", {})
        attrs      = session.get("attributes", {})

        row = {
            "timestamp":          event.get("eventTime"),
            "event_name":         event_name,
            "event_source":       event.get("eventSource"),
            "aws_region":         event.get("awsRegion"),
            "source_ip":          event.get("sourceIPAddress"),
            "error_code":         event.get("errorCode"),
            "label":              1 if event_name in ATTACK_EVENTS else 0,
            "attack_technique":   ATTACK_EVENTS.get(event_name),
            # ── new fields ────────────────────────────────────────────────────
            "read_only":          event.get("readOnly"),
            "user_agent":         event.get("userAgent"),
            "access_key_id":      identity.get("accessKeyId"),
            "mfa_authenticated":  attrs.get("mfaAuthenticated"),
            "target_resource":    extract_target_resource(params),
            "request_params_raw": json.dumps(params) if params else None,
            # ─────────────────────────────────────────────────────────────────
            **principal,
        }
        all_events.append(row)

df_enriched = pd.DataFrame(all_events)
df_enriched["timestamp"] = pd.to_datetime(df_enriched["timestamp"])
df_enriched.sort_values("timestamp", inplace=True)
df_enriched.reset_index(drop=True, inplace=True)

print(f"Shape: {df_enriched.shape}")
print(f"\nNew columns coverage (non-null %):")
new_cols = ["read_only", "user_agent", "access_key_id", "mfa_authenticated", "target_resource"]
for col in new_cols:
    pct = df_enriched[col].notna().mean() * 100
    print(f"  {col:25s}: {pct:.1f}%")

print(f"\nLabel split:  {df_enriched['label'].value_counts().to_dict()}")
print(f"\nread_only distribution:\n{df_enriched['read_only'].value_counts()}")
print(f"\nSample attack rows with target_resource:")
print(df_enriched[df_enriched['label'] == 1][
    ['event_name', 'attack_technique', 'target_resource', 'user_agent']
].drop_duplicates().head(15).to_string())


Shape: (2900, 17)

New columns coverage (non-null %):
  read_only                : 100.0%
  user_agent               : 100.0%
  access_key_id            : 97.1%
  mfa_authenticated        : 23.2%
  target_resource          : 61.2%

Label split:  {0: 2764, 1: 136}

read_only distribution:
read_only
True     2326
False     574
Name: count, dtype: int64

Sample attack rows with target_resource:
          event_name      attack_technique                              target_resource                                                                                                                                                                                                                                                                                   user_agent
87        CreateRole           persistence  stratus-red-team-ec2-get-password-data-role  APN/1.0 HashiCorp/1.0 Terraform/1.1.2 (+https://www.terraform.io) terraform-provider-aws/3.76.1 (+https://registry.terraform.io/providers/hashi

In [24]:
# ── Session labeling (same ±5 min window as before) ──────────────────────────
WINDOW_MINUTES = 5
df_enriched["session_label"] = df_enriched["label"].copy()

for username, group in df_enriched.groupby("username"):
    attack_times = group[group["label"] == 1]["timestamp"]
    for atk_time in attack_times:
        mask = (
            (df_enriched["username"] == username) &
            (df_enriched["timestamp"] >= atk_time - pd.Timedelta(minutes=WINDOW_MINUTES)) &
            (df_enriched["timestamp"] <= atk_time + pd.Timedelta(minutes=WINDOW_MINUTES))
        )
        df_enriched.loc[mask, "session_label"] = 1

# ── Save ───────────────────────────────────────────────────────────────────────
df_enriched.to_csv("invictus_enriched.csv", index=False)
print("Saved invictus_enriched.csv")
print(f"Shape:   {df_enriched.shape}")
print(f"Columns: {list(df_enriched.columns)}")
print(f"\nEvent-level attacks:   {df_enriched['label'].sum()}")
print(f"Session-level attacks: {df_enriched['session_label'].sum()}")
print(f"\nNull counts:\n{df_enriched.isnull().sum().to_string()}")


Saved invictus_enriched.csv
Shape:   (2900, 18)
Columns: ['timestamp', 'event_name', 'event_source', 'aws_region', 'source_ip', 'error_code', 'label', 'attack_technique', 'read_only', 'user_agent', 'access_key_id', 'mfa_authenticated', 'target_resource', 'request_params_raw', 'principal_type', 'principal_arn', 'username', 'session_label']

Event-level attacks:   136
Session-level attacks: 2597

Null counts:
timestamp                0
event_name               0
event_source             0
aws_region               0
source_ip                0
error_code            2600
label                    0
attack_technique      2764
read_only                0
user_agent               0
access_key_id           84
mfa_authenticated     2226
target_resource       1125
request_params_raw     333
principal_type           0
principal_arn           77
username                42
session_label            0


## Synthetic Attack Chain Generator

Generates CloudTrail-format attack sessions grounded in documented MITRE ATT&CK for Cloud techniques.
Each session randomises: username, account ID, resource names, source IP, timing jitter, tool user-agent,
and the amount of benign recon noise before/during/after the attack.

**Why this is defensible:**
- Attack chain *structure* comes from documented Rhino Security Labs / MITRE ATT&CK techniques — not invented
- Only surface attributes (names, IPs, timing) are randomised
- Results are always reported separately from real data, never merged silently
- Statistical similarity to real data is verified in the cell below

In [25]:
import random
import string
import json
import pandas as pd
from datetime import datetime, timezone, timedelta

# ── Reproducibility ────────────────────────────────────────────────────────────
random.seed(42)

# ─────────────────────────────────────────────────────────────────────────────
# ATTACK CHAIN LIBRARY
# Each chain is a list of steps. Each step is a dict:
#   event_name, event_source, attack_technique, read_only,
#   target_key (which field in requestParameters holds the resource name)
#   optional error_probability (0.0–1.0, default 0)
#
# Sources: Rhino Security Labs "AWS IAM Privilege Escalation Methods",
#          MITRE ATT&CK for Cloud T1078/T1098/T1136/T1484,
#          Stratus Red Team technique catalogue
# ─────────────────────────────────────────────────────────────────────────────

ATTACK_CHAINS = {

    # ── Chain 1: Create role + attach managed admin policy ────────────────────
    "create_role_attach_managed_policy": [
        {"event_name": "CreateRole",        "event_source": "iam.amazonaws.com", "attack_technique": "persistence",            "read_only": False, "target_key": "role"},
        {"event_name": "AttachRolePolicy",  "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation",   "read_only": False, "target_key": "role"},
    ],

    # ── Chain 2: Create role + inline admin policy ─────────────────────────────
    "create_role_inline_policy": [
        {"event_name": "CreateRole",        "event_source": "iam.amazonaws.com", "attack_technique": "persistence",            "read_only": False, "target_key": "role"},
        {"event_name": "PutRolePolicy",     "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation",   "read_only": False, "target_key": "role"},
    ],

    # ── Chain 3: Create user → access key → attach admin policy ───────────────
    "create_user_accesskey_policy": [
        {"event_name": "CreateUser",        "event_source": "iam.amazonaws.com", "attack_technique": "persistence",            "read_only": False, "target_key": "user"},
        {"event_name": "CreateAccessKey",   "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation",   "read_only": False, "target_key": "user"},
        {"event_name": "AttachUserPolicy",  "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation",   "read_only": False, "target_key": "user"},
    ],

    # ── Chain 4: Create user → console login profile ───────────────────────────
    "create_user_console_access": [
        {"event_name": "CreateUser",          "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "user"},
        {"event_name": "CreateLoginProfile",  "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "user"},
        {"event_name": "AttachUserPolicy",    "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "user"},
    ],

    # ── Chain 5: Add existing user to admin group ─────────────────────────────
    "add_user_to_admin_group": [
        {"event_name": "AddUserToGroup",    "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation",   "read_only": False, "target_key": "group"},
    ],

    # ── Chain 6: Update existing role's inline policy to admin ────────────────
    "update_role_inline_policy": [
        {"event_name": "PutRolePolicy",     "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation",   "read_only": False, "target_key": "role"},
    ],

    # ── Chain 7: Create new policy version with admin rights ──────────────────
    "create_policy_version": [
        {"event_name": "CreatePolicyVersion",    "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "policy"},
        {"event_name": "SetDefaultPolicyVersion","event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "policy"},
    ],

    # ── Chain 8: Update assume-role trust policy to allow attacker account ─────
    "update_assume_role_policy": [
        {"event_name": "UpdateAssumeRolePolicy", "event_source": "iam.amazonaws.com", "attack_technique": "persistence",         "read_only": False, "target_key": "role"},
        {"event_name": "AssumeRole",             "event_source": "sts.amazonaws.com",  "attack_technique": "privilege-escalation","read_only": False, "target_key": "role"},
    ],

    # ── Chain 9: Full kill chain (like bert-jan) ───────────────────────────────
    "full_kill_chain": [
        {"event_name": "CreateRole",         "event_source": "iam.amazonaws.com",           "attack_technique": "persistence",          "read_only": False, "target_key": "role"},
        {"event_name": "AttachRolePolicy",   "event_source": "iam.amazonaws.com",           "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
        {"event_name": "GetSecretValue",     "event_source": "secretsmanager.amazonaws.com","attack_technique": "credential-access",    "read_only": True,  "target_key": "secret"},
        {"event_name": "PutBucketPolicy",    "event_source": "s3.amazonaws.com",            "attack_technique": "exfiltration",         "read_only": False, "target_key": "bucket"},
        {"event_name": "StopLogging",        "event_source": "cloudtrail.amazonaws.com",    "attack_technique": "defense-evasion",      "read_only": False, "target_key": "trail",
         "error_probability": 0.4},  # often fails if trail name wrong
    ],

    # ── Chain 10: Credential access via EC2 password data ─────────────────────
    "ec2_password_data": [
        {"event_name": "CreateRole",        "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "role"},
        {"event_name": "PutRolePolicy",     "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
        {"event_name": "GetPasswordData",   "event_source": "ec2.amazonaws.com", "attack_technique": "credential-access",    "read_only": True,  "target_key": "instance",
         "error_probability": 0.9},  # usually fails unless Windows instance exists
    ],
}

# ── Recon events that typically precede an attack ─────────────────────────────
RECON_EVENTS = [
    ("GetAccountSummary",            "iam.amazonaws.com",           True),
    ("ListUsers",                    "iam.amazonaws.com",           True),
    ("ListRoles",                    "iam.amazonaws.com",           True),
    ("ListGroups",                   "iam.amazonaws.com",           True),
    ("ListPolicies",                 "iam.amazonaws.com",           True),
    ("GetAccountAuthorizationDetails","iam.amazonaws.com",          True),
    ("ListAttachedUserPolicies",     "iam.amazonaws.com",           True),
    ("ListAttachedRolePolicies",     "iam.amazonaws.com",           True),
    ("ListBuckets",                  "s3.amazonaws.com",            True),
    ("DescribeInstances",            "ec2.amazonaws.com",           True),
    ("ListSecrets",                  "secretsmanager.amazonaws.com",True),
    ("DescribeTrails",               "cloudtrail.amazonaws.com",    True),
    ("GetCallerIdentity",            "sts.amazonaws.com",           True),
    ("ListAccessKeys",               "iam.amazonaws.com",           True),
]

# ── Benign background events (non-attacker users) ─────────────────────────────
BENIGN_EVENTS = [
    ("GetBucketLogging",              "s3.amazonaws.com",            True),
    ("GetBucketPolicy",               "s3.amazonaws.com",            True),
    ("GetBucketAcl",                  "s3.amazonaws.com",            True),
    ("DescribeSecurityGroups",        "ec2.amazonaws.com",           True),
    ("DescribeVpcs",                  "ec2.amazonaws.com",           True),
    ("DescribeSubnets",               "ec2.amazonaws.com",           True),
    ("GetRegionOptStatus",            "account.amazonaws.com",       True),
    ("DescribeDBInstances",           "rds.amazonaws.com",           True),
    ("ListKeys",                      "kms.amazonaws.com",           True),
    ("DescribeKey",                   "kms.amazonaws.com",           True),
    ("GetParameter",                  "ssm.amazonaws.com",           True),
    ("DescribeInstanceInformation",   "ssm.amazonaws.com",           True),
    ("GetSecretValue",                "secretsmanager.amazonaws.com",True),  # benign app usage
    ("DescribeInstances",             "ec2.amazonaws.com",           True),
    ("ListFunctions",                 "lambda.amazonaws.com",        True),
]

# ── Realistic user agents ──────────────────────────────────────────────────────
USER_AGENTS = [
    "aws-cli/2.13.0 Python/3.11.4 Linux/5.15.0 botocore/2.0.0",
    "Boto3/1.28.0 Python/3.10.6 Linux/5.19.0 Botocore/1.31.0",
    "Boto3/1.26.165 Python/3.10.6 Linux/5.19.0-46-generic Botocore/1.29.165",
    "aws-cli/1.29.0 Python/3.9.0 Darwin/22.0.0 botocore/1.31.0",
    "Terraform/1.5.0 aws-sdk-go/1.44.300",
    "console.amazonaws.com",
    "AWS Internal",
]

ATTACKER_USER_AGENTS = [
    "aws-cli/2.13.0 Python/3.11.4 Linux/5.15.0 botocore/2.0.0",
    "Boto3/1.28.0 Python/3.10.6 Linux/5.19.0 Botocore/1.31.0",
    "python-requests/2.28.0",
]

ERROR_CODES = [
    "AccessDenied", "NoSuchEntity", "ThrottlingException",
    "InvalidParameterValue", "EntityAlreadyExists",
]

# ── Random value generators ────────────────────────────────────────────────────

def rand_str(n=8):
    return ''.join(random.choices(string.ascii_lowercase, k=n))

def rand_account_id():
    return ''.join(random.choices(string.digits, k=12))

def rand_ip():
    return f"{random.randint(10,203)}.{random.randint(0,255)}.{random.randint(0,255)}.{random.randint(1,254)}"

def rand_access_key(kind="long_term"):
    prefix = "AKIA" if kind == "long_term" else "ASIA"
    return prefix + ''.join(random.choices(string.ascii_uppercase + string.digits, k=16))

def rand_resource(kind):
    suffix = rand_str(6)
    mapping = {
        "role":    f"role-{suffix}",
        "user":    f"svc-{suffix}",
        "group":   f"admins-{suffix}",
        "policy":  f"arn:aws:iam::aws:policy/AdministratorAccess",
        "secret":  f"prod/db/{suffix}",
        "bucket":  f"data-{suffix}-bucket",
        "trail":   f"mgmt-trail-{suffix}",
        "instance":f"i-{rand_str(17)}",
    }
    return mapping.get(kind, suffix)

def jitter(seconds_range=(2, 45)):
    return timedelta(seconds=random.randint(*seconds_range))


# ── Core session generator ─────────────────────────────────────────────────────

def generate_session(
    chain_name: str,
    base_time: datetime = None,
    recon_events: int = 5,
    noise_events: int = 3,
    account_id: str = None,
) -> list[dict]:
    """
    Generate one synthetic session: recon → attack chain → post-exploitation noise.
    Returns a list of row dicts matching invictus_enriched.csv schema.
    """
    chain = ATTACK_CHAINS[chain_name]
    account_id = account_id or rand_account_id()
    attacker   = rand_str(7)
    source_ip  = rand_ip()
    access_key = rand_access_key("long_term")
    user_agent = random.choice(ATTACKER_USER_AGENTS)
    base_time  = base_time or datetime(2024, 1, random.randint(1, 28),
                                        random.randint(8, 18), 0, 0, tzinfo=timezone.utc)

    # Pre-generate resource names per target_key type needed by this chain
    resources = {}
    for step in chain:
        tk = step["target_key"]
        if tk not in resources:
            resources[tk] = rand_resource(tk)

    rows = []
    t = base_time

    # 1. Recon phase
    recon_pool = random.sample(RECON_EVENTS, min(recon_events, len(RECON_EVENTS)))
    for name, source, read_only in recon_pool:
        t += jitter((3, 30))
        rows.append({
            "timestamp":          t.isoformat(),
            "event_name":         name,
            "event_source":       source,
            "aws_region":         "us-east-1",
            "source_ip":          source_ip,
            "error_code":         None,
            "label":              0,               # recon is benign label (per-event)
            "attack_technique":   None,
            "read_only":          read_only,
            "user_agent":         user_agent,
            "access_key_id":      access_key,
            "mfa_authenticated":  random.choice(["true", "false"]),
            "target_resource":    None,
            "request_params_raw": None,
            "principal_type":     "IAMUser",
            "principal_arn":      f"arn:aws:iam::{account_id}:user/{attacker}",
            "username":           attacker,
            "session_label":      1,               # session IS malicious (attacker is active)
            "synthetic":          True,
        })

    # 2. Attack chain
    for step in chain:
        t += jitter((5, 60))
        error_prob  = step.get("error_probability", 0.0)
        error_code  = random.choice(ERROR_CODES) if random.random() < error_prob else None
        target_res  = resources[step["target_key"]]

        rows.append({
            "timestamp":          t.isoformat(),
            "event_name":         step["event_name"],
            "event_source":       step["event_source"],
            "aws_region":         "us-east-1",
            "source_ip":          source_ip,
            "error_code":         error_code,
            "label":              1,
            "attack_technique":   step["attack_technique"],
            "read_only":          step["read_only"],
            "user_agent":         user_agent,
            "access_key_id":      access_key,
            "mfa_authenticated":  "false",         # attackers rarely have MFA
            "target_resource":    target_res,
            "request_params_raw": json.dumps({step["target_key"] + "Name": target_res}),
            "principal_type":     "IAMUser",
            "principal_arn":      f"arn:aws:iam::{account_id}:user/{attacker}",
            "username":           attacker,
            "session_label":      1,
            "synthetic":          True,
        })

    # 3. Post-attack noise (makes session look more realistic)
    noise_pool = random.sample(BENIGN_EVENTS, min(noise_events, len(BENIGN_EVENTS)))
    for name, source, read_only in noise_pool:
        t += jitter((10, 90))
        rows.append({
            "timestamp":          t.isoformat(),
            "event_name":         name,
            "event_source":       source,
            "aws_region":         "us-east-1",
            "source_ip":          source_ip,
            "error_code":         None,
            "label":              0,
            "attack_technique":   None,
            "read_only":          read_only,
            "user_agent":         user_agent,
            "access_key_id":      access_key,
            "mfa_authenticated":  "false",
            "target_resource":    None,
            "request_params_raw": None,
            "principal_type":     "IAMUser",
            "principal_arn":      f"arn:aws:iam::{account_id}:user/{attacker}",
            "username":           attacker,
            "session_label":      1,
            "synthetic":          True,
        })

    return rows


def generate_benign_session(
    n_events: int = 15,
    base_time: datetime = None,
    account_id: str = None,
) -> list[dict]:
    """Generate a purely benign user session for negative class examples."""
    account_id = account_id or rand_account_id()
    username   = rand_str(6)
    source_ip  = rand_ip()
    access_key = rand_access_key("long_term")
    user_agent = random.choice(USER_AGENTS)
    base_time  = base_time or datetime(2024, 1, random.randint(1, 28),
                                        random.randint(8, 18), 0, 0, tzinfo=timezone.utc)
    rows = []
    t = base_time
    pool = random.choices(BENIGN_EVENTS, k=n_events)
    for name, source, read_only in pool:
        t += jitter((5, 120))
        error_code = random.choice(ERROR_CODES) if random.random() < 0.05 else None
        rows.append({
            "timestamp":          t.isoformat(),
            "event_name":         name,
            "event_source":       source,
            "aws_region":         "us-east-1",
            "source_ip":          source_ip,
            "error_code":         error_code,
            "label":              0,
            "attack_technique":   None,
            "read_only":          read_only,
            "user_agent":         user_agent,
            "access_key_id":      access_key,
            "mfa_authenticated":  random.choice(["true", "false"]),
            "target_resource":    None,
            "request_params_raw": None,
            "principal_type":     "IAMUser",
            "principal_arn":      f"arn:aws:iam::{account_id}:user/{username}",
            "username":           username,
            "session_label":      0,
            "synthetic":          True,
        })
    return rows


print("Attack chains available:")
for name in ATTACK_CHAINS:
    steps = [s["event_name"] for s in ATTACK_CHAINS[name]]
    print(f"  {name}: {' → '.join(steps)}")


Attack chains available:
  create_role_attach_managed_policy: CreateRole → AttachRolePolicy
  create_role_inline_policy: CreateRole → PutRolePolicy
  create_user_accesskey_policy: CreateUser → CreateAccessKey → AttachUserPolicy
  create_user_console_access: CreateUser → CreateLoginProfile → AttachUserPolicy
  add_user_to_admin_group: AddUserToGroup
  update_role_inline_policy: PutRolePolicy
  create_policy_version: CreatePolicyVersion → SetDefaultPolicyVersion
  update_assume_role_policy: UpdateAssumeRolePolicy → AssumeRole
  full_kill_chain: CreateRole → AttachRolePolicy → GetSecretValue → PutBucketPolicy → StopLogging
  ec2_password_data: CreateRole → PutRolePolicy → GetPasswordData


In [26]:
# ── Generate the synthetic dataset ────────────────────────────────────────────
# 20 attack sessions (2 per chain × 10 chains) + 40 benign sessions
# Gives ~800–1200 synthetic events; adjust N_PER_CHAIN to scale up

N_PER_CHAIN    = 10    # attack sessions generated per chain type
N_BENIGN       = 200  # purely benign sessions
RECON_RANGE    = (3, 8)   # random recon events per session
NOISE_RANGE    = (2, 5)   # random post-attack noise events per session

all_rows = []

# Attack sessions
for chain_name in ATTACK_CHAINS:
    for _ in range(N_PER_CHAIN):
        rows = generate_session(
            chain_name,
            recon_events=random.randint(*RECON_RANGE),
            noise_events=random.randint(*NOISE_RANGE),
        )
        all_rows.extend(rows)

# Benign sessions
for _ in range(N_BENIGN):
    rows = generate_benign_session(n_events=random.randint(8, 20))
    all_rows.extend(rows)

df_syn = pd.DataFrame(all_rows)
df_syn["timestamp"] = pd.to_datetime(df_syn["timestamp"])
df_syn.sort_values("timestamp", inplace=True)
df_syn.reset_index(drop=True, inplace=True)

print(f"Synthetic dataset shape:  {df_syn.shape}")
print(f"Attack sessions:          {N_PER_CHAIN * len(ATTACK_CHAINS)}")
print(f"Benign sessions:          {N_BENIGN}")
print(f"\nLabel split (event-level):")
print(f"  Benign  (0): {(df_syn['label']==0).sum()}")
print(f"  Attack  (1): {(df_syn['label']==1).sum()}")
print(f"\nSession label split:")
print(f"  Benign  (0): {(df_syn['session_label']==0).sum()}")
print(f"  Attack  (1): {(df_syn['session_label']==1).sum()}")
print(f"\nAttack technique breakdown:")
print(df_syn[df_syn['label']==1]['attack_technique'].value_counts().to_string())
print(f"\nChain diversity (unique attack event_name combos per session):")
attack_sessions = df_syn[df_syn['label']==1].groupby('username')['event_name'].apply(
    lambda x: tuple(x.tolist())
)
print(f"  {len(attack_sessions.unique())} unique attack sequences across {len(attack_sessions)} attack sessions")

# Save
df_syn.to_csv("synthetic_cloudtrail_extend.csv", index=False)
print("\nSaved synthetic_cloudtrail_extend.csv")


Synthetic dataset shape:  (3879, 19)
Attack sessions:          100
Benign sessions:          200

Label split (event-level):
  Benign  (0): 3639
  Attack  (1): 240

Session label split:
  Benign  (0): 2724
  Attack  (1): 1155

Attack technique breakdown:
attack_technique
privilege-escalation    130
persistence              70
credential-access        20
exfiltration             10
defense-evasion          10

Chain diversity (unique attack event_name combos per session):
  10 unique attack sequences across 100 attack sessions

Saved synthetic_cloudtrail_extend.csv


In [27]:
# ── Statistical similarity validation ─────────────────────────────────────────
# Compare synthetic vs real data across key distributions.
# This is what you present to justify synthetic data in the paper.

df_real = pd.read_csv("invictus_enriched.csv")
df_real["synthetic"] = False

print("=" * 55)
print("SYNTHETIC vs REAL — STATISTICAL SIMILARITY CHECK")
print("=" * 55)

# 1. read_only ratio
real_ro  = df_real["read_only"].mean()
syn_ro   = df_syn["read_only"].mean()
print(f"\nread_only ratio (write ops = False):")
print(f"  Real:      {real_ro:.3f}  ({(~df_real['read_only']).sum()} write events)")
print(f"  Synthetic: {syn_ro:.3f}  ({(~df_syn['read_only']).sum()} write events)")

# 2. error rate
real_err  = df_real["error_code"].notna().mean()
syn_err   = df_syn["error_code"].notna().mean()
print(f"\nError rate:")
print(f"  Real:      {real_err:.3f}")
print(f"  Synthetic: {syn_err:.3f}")

# 3. attack event name overlap
real_attack_events = set(df_real[df_real["label"]==1]["event_name"].unique())
syn_attack_events  = set(df_syn[df_syn["label"]==1]["event_name"].unique())
overlap = real_attack_events & syn_attack_events
print(f"\nAttack event_name overlap:")
print(f"  Real unique:      {real_attack_events}")
print(f"  Synthetic unique: {syn_attack_events}")
print(f"  Overlap:          {overlap}")

# 4. IAM event proportion
real_iam = (df_real["event_source"] == "iam.amazonaws.com").mean()
syn_iam  = (df_syn["event_source"]  == "iam.amazonaws.com").mean()
print(f"\nIAM event proportion:")
print(f"  Real:      {real_iam:.3f}")
print(f"  Synthetic: {syn_iam:.3f}")

# 5. principal_type distribution
print(f"\nprincipal_type distribution:")
print(f"  Real:\n{df_real['principal_type'].value_counts(normalize=True).round(3).to_string()}")
print(f"  Synthetic:\n{df_syn['principal_type'].value_counts(normalize=True).round(3).to_string()}")

# 6. mfa_authenticated in attack sessions
real_mfa_atk = df_real[df_real["label"]==1]["mfa_authenticated"].value_counts()
syn_mfa_atk  = df_syn[df_syn["label"]==1]["mfa_authenticated"].value_counts()
print(f"\nmfa_authenticated for attack events:")
print(f"  Real:      {real_mfa_atk.to_dict()}")
print(f"  Synthetic: {syn_mfa_atk.to_dict()}")

print(f"\n{'='*55}")
print("NOTE: synthetic data covers 10 distinct attack chain types")
print("vs 1 real attacker chain. Technique diversity is the goal,")
print("not pixel-perfect distribution matching.")


SYNTHETIC vs REAL — STATISTICAL SIMILARITY CHECK

read_only ratio (write ops = False):
  Real:      0.802  (574 write events)
  Synthetic: 0.943  (220 write events)

Error rate:
  Real:      0.103
  Synthetic: 0.039

Attack event_name overlap:
  Real unique:      {'AttachUserPolicy', 'DeleteTrail', 'AttachRolePolicy', 'CreateRole', 'GetPasswordData', 'GetSecretValue', 'PutBucketPolicy', 'StopLogging', 'DeleteBucketPolicy', 'UpdateAssumeRolePolicy', 'CreateUser', 'PutRolePolicy', 'CreateAccessKey', 'PutEventSelectors', 'CreateLoginProfile'}
  Synthetic unique: {'AttachUserPolicy', 'CreatePolicyVersion', 'AssumeRole', 'AttachRolePolicy', 'CreateRole', 'PutRolePolicy', 'GetSecretValue', 'PutBucketPolicy', 'GetPasswordData', 'UpdateAssumeRolePolicy', 'CreateLoginProfile', 'SetDefaultPolicyVersion', 'StopLogging', 'CreateUser', 'CreateAccessKey', 'AddUserToGroup'}
  Overlap:          {'AttachUserPolicy', 'AttachRolePolicy', 'CreateRole', 'GetPasswordData', 'GetSecretValue', 'PutBucketPolicy

## Synthetic Data — Gap Fix

Fixes 3 statistical gaps found by validation against real invictus data:

| Gap | Real | Old Synthetic | Fix |
|---|---|---|---|
| Error rate | 10.3% | 3.9% | Raise benign error prob to 12%, add ThrottlingException |
| read_only ratio | 80.2% | 94.3% | Add benign write-ops (SSM puts, EC2 actions, Secrets rotation) |
| Principal type diversity | IAMUser 94.8%, AssumedRole 2.6%, AWSService 1.2% | 100% IAMUser | Generate 15% of benign sessions as AssumedRole |

In [30]:
import random, string, json, pandas as pd
from datetime import datetime, timezone, timedelta

random.seed(99)

# ── Unchanged attack chain library (same as before) ───────────────────────────
ATTACK_CHAINS = {
    "create_role_attach_managed_policy": [
        {"event_name": "CreateRole",        "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "role"},
        {"event_name": "AttachRolePolicy",  "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
    ],
    "create_role_inline_policy": [
        {"event_name": "CreateRole",    "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "role"},
        {"event_name": "PutRolePolicy", "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
    ],
    "create_user_accesskey_policy": [
        {"event_name": "CreateUser",       "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "user"},
        {"event_name": "CreateAccessKey",  "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "user"},
        {"event_name": "AttachUserPolicy", "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "user"},
    ],
    "create_user_console_access": [
        {"event_name": "CreateUser",         "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "user"},
        {"event_name": "CreateLoginProfile", "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "user"},
        {"event_name": "AttachUserPolicy",   "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "user"},
    ],
    "add_user_to_admin_group": [
        {"event_name": "AddUserToGroup", "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "group"},
    ],
    "update_role_inline_policy": [
        {"event_name": "PutRolePolicy", "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
    ],
    "create_policy_version": [
        {"event_name": "CreatePolicyVersion",     "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "policy"},
        {"event_name": "SetDefaultPolicyVersion", "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "policy"},
    ],
    "update_assume_role_policy": [
        {"event_name": "UpdateAssumeRolePolicy", "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "role"},
        {"event_name": "AssumeRole",             "event_source": "sts.amazonaws.com",  "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
    ],
    "full_kill_chain": [
        {"event_name": "CreateRole",       "event_source": "iam.amazonaws.com",            "attack_technique": "persistence",          "read_only": False, "target_key": "role"},
        {"event_name": "AttachRolePolicy", "event_source": "iam.amazonaws.com",            "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
        {"event_name": "GetSecretValue",   "event_source": "secretsmanager.amazonaws.com", "attack_technique": "credential-access",    "read_only": True,  "target_key": "secret"},
        {"event_name": "PutBucketPolicy",  "event_source": "s3.amazonaws.com",             "attack_technique": "exfiltration",         "read_only": False, "target_key": "bucket"},
        {"event_name": "StopLogging",      "event_source": "cloudtrail.amazonaws.com",     "attack_technique": "defense-evasion",      "read_only": False, "target_key": "trail", "error_probability": 0.4},
    ],
    "ec2_password_data": [
        {"event_name": "CreateRole",      "event_source": "iam.amazonaws.com", "attack_technique": "persistence",          "read_only": False, "target_key": "role"},
        {"event_name": "PutRolePolicy",   "event_source": "iam.amazonaws.com", "attack_technique": "privilege-escalation", "read_only": False, "target_key": "role"},
        {"event_name": "GetPasswordData", "event_source": "ec2.amazonaws.com", "attack_technique": "credential-access",    "read_only": True,  "target_key": "instance", "error_probability": 0.9},
    ],
}

RECON_EVENTS = [
    ("GetAccountSummary",             "iam.amazonaws.com",            True),
    ("ListUsers",                     "iam.amazonaws.com",            True),
    ("ListRoles",                     "iam.amazonaws.com",            True),
    ("ListGroups",                    "iam.amazonaws.com",            True),
    ("ListPolicies",                  "iam.amazonaws.com",            True),
    ("GetAccountAuthorizationDetails","iam.amazonaws.com",            True),
    ("ListAttachedUserPolicies",      "iam.amazonaws.com",            True),
    ("ListAttachedRolePolicies",      "iam.amazonaws.com",            True),
    ("ListBuckets",                   "s3.amazonaws.com",             True),
    ("DescribeInstances",             "ec2.amazonaws.com",            True),
    ("ListSecrets",                   "secretsmanager.amazonaws.com", True),
    ("DescribeTrails",                "cloudtrail.amazonaws.com",     True),
    ("GetCallerIdentity",             "sts.amazonaws.com",            True),
    ("ListAccessKeys",                "iam.amazonaws.com",            True),
]

# ── FIX 1 & 2: Expanded benign event pool — includes write ops ────────────────
# Weighted: (event_name, event_source, read_only, weight)
# write-op weight tuned so overall read_only ratio hits ~0.80 matching real data
BENIGN_EVENTS_WEIGHTED = [
    # Read-only (common background noise)
    ("GetBucketLogging",              "s3.amazonaws.com",             True,  4),
    ("GetBucketPolicy",               "s3.amazonaws.com",             True,  4),
    ("GetBucketAcl",                  "s3.amazonaws.com",             True,  3),
    ("DescribeSecurityGroups",        "ec2.amazonaws.com",            True,  5),
    ("DescribeVpcs",                  "ec2.amazonaws.com",            True,  4),
    ("DescribeSubnets",               "ec2.amazonaws.com",            True,  4),
    ("DescribeInstances",             "ec2.amazonaws.com",            True,  5),
    ("GetRegionOptStatus",            "account.amazonaws.com",        True,  2),
    ("DescribeDBInstances",           "rds.amazonaws.com",            True,  3),
    ("ListKeys",                      "kms.amazonaws.com",            True,  4),
    ("DescribeKey",                   "kms.amazonaws.com",            True,  3),
    ("GetParameter",                  "ssm.amazonaws.com",            True,  5),
    ("DescribeInstanceInformation",   "ssm.amazonaws.com",            True,  4),
    ("GetSecretValue",                "secretsmanager.amazonaws.com", True,  4),
    ("ListFunctions",                 "lambda.amazonaws.com",         True,  2),
    ("DescribeLoadBalancers",         "elasticloadbalancing.amazonaws.com", True, 2),
    # IAM read events — benign users routinely inspect roles/users/policies
    # Breaks the pattern where role-xxx/svc-xxx names only appear in attack sessions
    ("GetRole",           "iam.amazonaws.com", True,  3),
    ("GetUser",           "iam.amazonaws.com", True,  3),
    ("ListRolePolicies",  "iam.amazonaws.com", True,  2),
    ("GetRolePolicy",     "iam.amazonaws.com", True,  2),
    ("GetUserPolicy",     "iam.amazonaws.com", True,  1),
    ("ListGroupsForUser", "iam.amazonaws.com", True,  1),
    # FIX 2 — Benign WRITE ops (present in real data, legitimately write=False)
    ("PutParameter",                  "ssm.amazonaws.com",            False, 3),  # SSM config writes
    ("SendCommand",                   "ssm.amazonaws.com",            False, 3),  # SSM run command
    ("StartInstances",                "ec2.amazonaws.com",            False, 2),  # EC2 lifecycle
    ("StopInstances",                 "ec2.amazonaws.com",            False, 2),  # EC2 lifecycle
    ("RebootInstances",               "ec2.amazonaws.com",            False, 1),
    ("ModifyInstanceAttribute",       "ec2.amazonaws.com",            False, 2),
    ("RotateSecret",                  "secretsmanager.amazonaws.com", False, 2),  # routine rotation
    ("PutSecretValue",                "secretsmanager.amazonaws.com", False, 2),  # app secret update
    ("CreateSnapshot",                "ec2.amazonaws.com",            False, 1),
    ("ModifyDBInstance",              "rds.amazonaws.com",            False, 1),
]

# FIX 3 — AssumedRole benign session template (for services/automation)
ASSUMED_ROLE_BENIGN = [
    ("DescribeInstances",   "ec2.amazonaws.com",            True,  5),
    ("GetParameter",        "ssm.amazonaws.com",            True,  5),
    ("GetSecretValue",      "secretsmanager.amazonaws.com", True,  4),
    ("ListKeys",            "kms.amazonaws.com",            True,  3),
    ("SendCommand",         "ssm.amazonaws.com",            False, 3),
    ("PutParameter",        "ssm.amazonaws.com",            False, 2),
    ("DescribeDBInstances", "rds.amazonaws.com",            True,  2),
]

# FIX 1 — Realistic error codes with correct weights
BENIGN_ERROR_CODES = [
    ("ThrottlingException",                  40),  # most common in real data
    ("Client.UnauthorizedOperation",         17),
    ("AccessDenied",                          6),
    ("NoSuchBucketPolicy",                    5),
    ("Client.InvalidRouteTableID.NotFound",   5),
    ("NoSuchPublicAccessBlockConfiguration",  5),
    ("NoSuchWebsiteConfiguration",            4),
    ("NoSuchCORSConfiguration",               4),
    ("NoSuchLifecycleConfiguration",          4),
]
_benign_err_pool  = [code for code, w in BENIGN_ERROR_CODES for _ in range(w)]
ATTACK_ERROR_CODES = ["AccessDenied", "NoSuchEntity", "ThrottlingException", "InvalidParameterValue"]

USER_AGENTS = [
    "aws-cli/2.13.0 Python/3.11.4 Linux/5.15.0 botocore/2.0.0",
    "Boto3/1.28.0 Python/3.10.6 Linux/5.19.0 Botocore/1.31.0",
    "Boto3/1.26.165 Python/3.10.6 Linux/5.19.0-46-generic Botocore/1.29.165",
    "aws-cli/1.29.0 Python/3.9.0 Darwin/22.0.0 botocore/1.31.0",
    "Terraform/1.5.0 aws-sdk-go/1.44.300",
    "console.amazonaws.com",
    "AWS Internal",
]
ATTACKER_UAS = [
    "aws-cli/2.13.0 Python/3.11.4 Linux/5.15.0 botocore/2.0.0",
    "Boto3/1.28.0 Python/3.10.6 Linux/5.19.0 Botocore/1.31.0",
    "python-requests/2.28.0",
]

def rand_str(n=8):   return "".join(random.choices(string.ascii_lowercase, k=n))
def rand_account():  return "".join(random.choices(string.digits, k=12))
def rand_ip():       return f"{random.randint(10,203)}.{random.randint(0,255)}.{random.randint(0,255)}.{random.randint(1,254)}"
def rand_key():      return "AKIA" + "".join(random.choices(string.ascii_uppercase + string.digits, k=16))
def rand_role_key(): return "ASIA" + "".join(random.choices(string.ascii_uppercase + string.digits, k=16))
def rand_resource(kind):
    s = rand_str(6)
    return {"role": f"role-{s}", "user": f"svc-{s}", "group": f"admins-{s}",
            "policy": "arn:aws:iam::aws:policy/AdministratorAccess",
            "secret": f"prod/db/{s}", "bucket": f"data-{s}-bucket",
            "trail": f"mgmt-trail-{s}", "instance": f"i-{rand_str(17)}"}.get(kind, s)
def jitter(lo=2, hi=45): return timedelta(seconds=random.randint(lo, hi))

def _weighted_sample(pool, n):
    """Sample n items from weighted pool [(item, weight), ...]"""
    items  = [x[:-1] if len(x)==4 else x for x in pool]
    weights= [x[-1] for x in pool]
    return random.choices(items, weights=weights, k=n)


# ── Attack session generator (unchanged logic, same as before) ─────────────────
def generate_attack_session(chain_name, recon_events=5, noise_events=3):
    chain      = ATTACK_CHAINS[chain_name]
    account_id = rand_account(); attacker = rand_str(random.randint(4, 12))
    source_ip  = rand_ip(); access_key = rand_key()
    ua         = random.choice(ATTACKER_UAS)
    t          = datetime(2024, random.randint(1,12), random.randint(1,28),
                          random.randint(7,19), 0, 0, tzinfo=timezone.utc)
    resources  = {s["target_key"]: rand_resource(s["target_key"]) for s in chain}
    rows = []

    for name, source, ro in random.sample(RECON_EVENTS, min(recon_events, len(RECON_EVENTS))):
        t += jitter(3, 30)
        rows.append({"timestamp": t.isoformat(), "event_name": name, "event_source": source,
            "aws_region": "us-east-1", "source_ip": source_ip, "error_code": None,
            "label": 0, "attack_technique": None, "read_only": ro, "user_agent": ua,
            "access_key_id": access_key, "mfa_authenticated": random.choice(["true","false"]),
            "target_resource": None, "request_params_raw": None, "principal_type": "IAMUser",
            "principal_arn": f"arn:aws:iam::{account_id}:user/{attacker}",
            "username": attacker, "session_label": 1, "synthetic": True})

    for step in chain:
        t += jitter(5, 60)
        ep = step.get("error_probability", 0)
        ec = random.choice(ATTACK_ERROR_CODES) if random.random() < ep else None
        rows.append({"timestamp": t.isoformat(), "event_name": step["event_name"],
            "event_source": step["event_source"], "aws_region": "us-east-1",
            "source_ip": source_ip, "error_code": ec, "label": 1,
            "attack_technique": step["attack_technique"], "read_only": step["read_only"],
            "user_agent": ua, "access_key_id": access_key, "mfa_authenticated": "false",
            "target_resource": resources[step["target_key"]],
            "request_params_raw": json.dumps({step["target_key"]+"Name": resources[step["target_key"]]}),
            "principal_type": "IAMUser",
            "principal_arn": f"arn:aws:iam::{account_id}:user/{attacker}",
            "username": attacker, "session_label": 1, "synthetic": True})

    for name, source, ro in _weighted_sample(BENIGN_EVENTS_WEIGHTED, noise_events):
        t += jitter(10, 90)
        rows.append({"timestamp": t.isoformat(), "event_name": name, "event_source": source,
            "aws_region": "us-east-1", "source_ip": source_ip, "error_code": None,
            "label": 0, "attack_technique": None, "read_only": ro, "user_agent": ua,
            "access_key_id": access_key, "mfa_authenticated": "false",
            "target_resource": None, "request_params_raw": None, "principal_type": "IAMUser",
            "principal_arn": f"arn:aws:iam::{account_id}:user/{attacker}",
            "username": attacker, "session_label": 1, "synthetic": True})
    return rows


# ── FIX 1+2: Patched IAMUser benign session ────────────────────────────────────
def generate_benign_iamuser(n_events=15):
    account_id = rand_account(); username = rand_str(6)
    source_ip  = rand_ip(); access_key = rand_key()
    ua         = random.choice(USER_AGENTS)
    t          = datetime(2024, random.randint(1,12), random.randint(1,28),
                          random.randint(7,19), 0, 0, tzinfo=timezone.utc)
    rows = []
    for name, source, ro in _weighted_sample(BENIGN_EVENTS_WEIGHTED, n_events):
        t += jitter(5, 120)
        # FIX 1: error rate 12% (up from 5%), weighted toward ThrottlingException
        ec = random.choice(_benign_err_pool) if random.random() < 0.12 else None
        rows.append({"timestamp": t.isoformat(), "event_name": name, "event_source": source,
            "aws_region": "us-east-1", "source_ip": source_ip, "error_code": ec,
            "label": 0, "attack_technique": None, "read_only": ro, "user_agent": ua,
            "access_key_id": access_key, "mfa_authenticated": random.choice(["true","false"]),
            "target_resource": None, "request_params_raw": None, "principal_type": "IAMUser",
            "principal_arn": f"arn:aws:iam::{account_id}:user/{username}",
            "username": username, "session_label": 0, "synthetic": True})
    return rows


# ── FIX 3: AssumedRole benign session (automation/service accounts) ────────────
def generate_benign_assumed_role(n_events=12):
    account_id = rand_account()
    role_name  = random.choice(["AWSServiceRoleForEC2", "LambdaExecutionRole",
                                 "ECSTaskRole", "CodeDeployRole", "AutoScalingRole"])
    session_id = rand_str(16)
    source_ip  = random.choice(["AWS Internal", rand_ip()])
    access_key = rand_role_key()
    ua         = random.choice(["AWS Internal", "aws-sdk-java/1.11.x", "aws-sdk-go/1.44.x"])
    t          = datetime(2024, random.randint(1,12), random.randint(1,28),
                          random.randint(7,19), 0, 0, tzinfo=timezone.utc)
    arn        = f"arn:aws:sts::{account_id}:assumed-role/{role_name}/{session_id}"
    rows = []
    for name, source, ro in _weighted_sample(ASSUMED_ROLE_BENIGN, n_events):
        t += jitter(1, 30)
        ec = random.choice(_benign_err_pool) if random.random() < 0.08 else None
        rows.append({"timestamp": t.isoformat(), "event_name": name, "event_source": source,
            "aws_region": "us-east-1", "source_ip": source_ip, "error_code": ec,
            "label": 0, "attack_technique": None, "read_only": ro, "user_agent": ua,
            "access_key_id": access_key, "mfa_authenticated": "false",
            "target_resource": None, "request_params_raw": None,
            "principal_type": "AssumedRole", "principal_arn": arn,
            "username": role_name, "session_label": 0, "synthetic": True})
    return rows


# ── Generate the fixed dataset ─────────────────────────────────────────────────
N_PER_CHAIN       = 20   # attack sessions per chain type  -> 200 attack sessions
N_BENIGN_IAMUSER  = 340  # ~85% of benign sessions
N_BENIGN_ASSUMED  = 60   # ~15% of benign sessions -> matches real 2.6% AssumedRole

all_rows = []

for chain_name in ATTACK_CHAINS:
    for _ in range(N_PER_CHAIN):
        all_rows.extend(generate_attack_session(
            chain_name,
            recon_events=random.randint(3, 8),
            noise_events=random.randint(2, 5),
        ))

for _ in range(N_BENIGN_IAMUSER):
    all_rows.extend(generate_benign_iamuser(n_events=random.randint(8, 20)))

for _ in range(N_BENIGN_ASSUMED):
    all_rows.extend(generate_benign_assumed_role(n_events=random.randint(6, 15)))

df_fixed = pd.DataFrame(all_rows)
df_fixed["timestamp"] = pd.to_datetime(df_fixed["timestamp"])
df_fixed.sort_values("timestamp", inplace=True)
df_fixed.reset_index(drop=True, inplace=True)

print(f"Shape: {df_fixed.shape}")
print(f"Attack sessions: {N_PER_CHAIN * len(ATTACK_CHAINS)}")
print(f"Benign sessions: {N_BENIGN_IAMUSER + N_BENIGN_ASSUMED}  "
      f"({N_BENIGN_IAMUSER} IAMUser + {N_BENIGN_ASSUMED} AssumedRole)")
print(f"\nLabel split:")
print(f"  Benign  (0): {(df_fixed['label']==0).sum()}")
print(f"  Attack  (1): {(df_fixed['label']==1).sum()}")


Shape: (3825, 19)
Attack sessions: 100
Benign sessions: 200  (170 IAMUser + 30 AssumedRole)

Label split:
  Benign  (0): 3585
  Attack  (1): 240


In [31]:
# ── Validate all 3 fixes against real data ────────────────────────────────────
df_real = pd.read_csv("invictus_enriched.csv")

print("=" * 55)
print("GAP VALIDATION — Fixed vs Real")
print("=" * 55)

checks = [
    ("read_only ratio",
     df_real["read_only"].mean(),
     df_fixed["read_only"].mean(),
     0.80, 0.02),
    ("error_code rate",
     df_real["error_code"].notna().mean(),
     df_fixed["error_code"].notna().mean(),
     0.10, 0.04),
]

for name, real_val, syn_val, target, tolerance in checks:
    gap    = abs(real_val - syn_val)
    status = "PASS" if gap <= tolerance else "NEEDS TUNING"
    print(f"\n{name}")
    print(f"  Real:      {real_val:.3f}")
    print(f"  Synthetic: {syn_val:.3f}  (gap={gap:.3f}, tolerance±{tolerance})  [{status}]")

print(f"\nprincipal_type distribution")
real_pt = df_real["principal_type"].value_counts(normalize=True).round(3)
syn_pt  = df_fixed["principal_type"].value_counts(normalize=True).round(3)
print(f"  Real:      {real_pt.to_dict()}")
print(f"  Synthetic: {syn_pt.to_dict()}")
assumed_pct = syn_pt.get("AssumedRole", 0)
status = "PASS" if assumed_pct > 0.05 else "LOW"
print(f"  AssumedRole coverage: {assumed_pct:.3f}  [{status}]")

print(f"\nerror_code distribution (top 5)")
real_ec = df_real["error_code"].value_counts(normalize=True).head(5).round(3)
syn_ec  = df_fixed["error_code"].value_counts(normalize=True).head(5).round(3)
print(f"  Real:      {real_ec.to_dict()}")
print(f"  Synthetic: {syn_ec.to_dict()}")

print(f"\nwrite op event sources (read_only=False)")
real_ws = df_real[~df_real["read_only"]]["event_source"].value_counts().head(5)
syn_ws  = df_fixed[~df_fixed["read_only"]]["event_source"].value_counts().head(5)
print(f"  Real:      {real_ws.to_dict()}")
print(f"  Synthetic: {syn_ws.to_dict()}")

# ── Save ──────────────────────────────────────────────────────────────────────
df_fixed.to_csv("synthetic_cloudtrail_fixed.csv", index=False)
print(f"\nSaved synthetic_cloudtrail_fixed.csv  ({df_fixed.shape[0]} rows)")
print("This replaces synthetic_cloudtrail_extend.csv as the training dataset.")


GAP VALIDATION — Fixed vs Real

read_only ratio
  Real:      0.802
  Synthetic: 0.752  (gap=0.050, tolerance±0.02)  [NEEDS TUNING]

error_code rate
  Real:      0.103
  Synthetic: 0.086  (gap=0.017, tolerance±0.04)  [PASS]

principal_type distribution
  Real:      {'IAMUser': 0.948, 'AssumedRole': 0.026, 'unknown': 0.014, 'AWSService': 0.012}
  Synthetic: {'IAMUser': 0.914, 'AssumedRole': 0.086}
  AssumedRole coverage: 0.086  [PASS]

error_code distribution (top 5)
  Real:      {'ThrottlingException': 0.34, 'Client.UnauthorizedOperation': 0.147, 'AccessDenied': 0.053, 'NoSuchBucketPolicy': 0.047, 'Client.InvalidRouteTableID.NotFound': 0.043}
  Synthetic: {'ThrottlingException': 0.456, 'Client.UnauthorizedOperation': 0.17, 'AccessDenied': 0.079, 'NoSuchLifecycleConfiguration': 0.058, 'NoSuchCORSConfiguration': 0.055}

write op event sources (read_only=False)
  Real:      {'ssm.amazonaws.com': 165, 'ec2.amazonaws.com': 155, 'secretsmanager.amazonaws.com': 97, 'iam.amazonaws.com': 88, 's3

## Baseline Evaluation — Rule-Based Detection

Establishes the rule-based baselines that the GNN+Sequence model must surpass.

**Evaluation protocol:**
- Train set: `synthetic_cloudtrail_extend.csv` (used to train GNN and sequence model)
- Test set: `invictus_enriched.csv` (real incident data, never used during training)
- Unit of evaluation: **session** (all events from one principal)
- Primary metric: F1 score at session level

In [29]:
import pandas as pd
from sklearn.metrics import precision_score, recall_score, f1_score, classification_report

# ── Rule sets ──────────────────────────────────────────────────────────────────
# Each represents a different real-world detection maturity level.
# Source: AWS GuardDuty findings catalogue, Rhino Security Labs IAM list.

RULES = {
    # What a junior analyst puts in a SIEM on day 1 — obvious defense evasion only
    "Minimal SIEM (3 rules)": {
        "StopLogging", "DeleteTrail", "CreateLoginProfile",
    },

    # Approximates GuardDuty IAM findings — covers well-known privilege escalation
    # but NOT AddUserToGroup, UpdateAssumeRolePolicy, or AssumeRole (all legitimate APIs)
    "GuardDuty-style (11 rules)": {
        "CreateLoginProfile", "UpdateLoginProfile",
        "AttachUserPolicy", "AttachRolePolicy", "AttachGroupPolicy",
        "PutUserPolicy", "PutRolePolicy",
        "CreatePolicyVersion", "SetDefaultPolicyVersion",
        "StopLogging", "DeleteTrail",
    },

    # Everything in the known-bad list — what you'd have AFTER reading the incident report
    # (represents post-incident rule tuning, not pre-incident detection)
    "Post-incident rules (all 23)": {
        "CreateAccessKey", "CreateLoginProfile", "UpdateLoginProfile",
        "AttachUserPolicy", "AttachRolePolicy", "AttachGroupPolicy",
        "PutUserPolicy", "PutRolePolicy", "PutGroupPolicy",
        "CreatePolicyVersion", "SetDefaultPolicyVersion", "AddUserToGroup",
        "CreateUser", "CreateRole", "UpdateAssumeRolePolicy",
        "StopLogging", "DeleteTrail", "UpdateTrail", "PutEventSelectors",
        "GetSecretValue", "GetPasswordData", "PutBucketPolicy", "DeleteBucketPolicy",
    },
}

def build_sessions(df, label_col="session_label"):
    """Aggregate events into sessions keyed by username."""
    sessions = {}
    for username, grp in df.groupby("username"):
        sessions[username] = {
            "events":     list(grp["event_name"]),
            "true_label": int(grp[label_col].max()),
        }
    return sessions

def rule_predict(sessions, rule_set):
    return {u: int(any(e in rule_set for e in s["events"])) for u, s in sessions.items()}

def report(sessions, preds, name):
    y_true = [sessions[u]["true_label"] for u in sessions]
    y_pred = [preds[u] for u in sessions]
    p = precision_score(y_true, y_pred, zero_division=0)
    r = recall_score(y_true, y_pred, zero_division=0)
    f = f1_score(y_true, y_pred, zero_division=0)
    tp = sum(1 for u in sessions if preds[u]==1 and sessions[u]["true_label"]==1)
    fp = sum(1 for u in sessions if preds[u]==1 and sessions[u]["true_label"]==0)
    fn = sum(1 for u in sessions if preds[u]==0 and sessions[u]["true_label"]==1)
    missed = [u for u in sessions if preds[u]==0 and sessions[u]["true_label"]==1]
    return {"name": name, "precision": p, "recall": r, "f1": f,
            "tp": tp, "fp": fp, "fn": fn, "missed": missed}

# ── Evaluate on SYNTHETIC (train distribution) ────────────────────────────────
print("=" * 60)
print("EVALUATION ON SYNTHETIC DATA (train distribution)")
print("=" * 60)
df_syn  = pd.read_csv("synthetic_cloudtrail_extend.csv")
sess_syn = build_sessions(df_syn)
print(f"Sessions: {len(sess_syn)}  |  "
      f"Attack: {sum(1 for s in sess_syn.values() if s['true_label']==1)}  |  "
      f"Benign: {sum(1 for s in sess_syn.values() if s['true_label']==0)}\n")

results_syn = []
for rule_name, rule_set in RULES.items():
    preds = rule_predict(sess_syn, rule_set)
    r = report(sess_syn, preds, rule_name)
    results_syn.append(r)
    print(f"{rule_name}")
    print(f"  P={r['precision']:.3f}  R={r['recall']:.3f}  F1={r['f1']:.3f}  "
          f"(TP={r['tp']} FP={r['fp']} FN={r['fn']})")
    if r["missed"]:
        # Show which chain types were missed
        missed_chains = set()
        for u in r["missed"]:
            evts = tuple(e for e in sess_syn[u]["events"]
                         if e not in {"GetAccountSummary","ListUsers","ListRoles",
                                       "ListGroups","ListPolicies","ListBuckets",
                                       "DescribeInstances","ListSecrets","DescribeTrails",
                                       "GetCallerIdentity","ListAccessKeys","GetBucketLogging",
                                       "DescribeSecurityGroups","DescribeVpcs","GetParameter",
                                       "DescribeDBInstances","ListKeys","DescribeKey",
                                       "ListFunctions","GetRegionOptStatus"})
            missed_chains.add(evts)
        print(f"  Missed attack patterns: {[' → '.join(c) for c in missed_chains]}")
    print()

# ── Evaluate on REAL invictus (test set — ground truth) ───────────────────────
print("=" * 60)
print("EVALUATION ON REAL INVICTUS DATA (test set)")
print("=" * 60)
df_real  = pd.read_csv("invictus_enriched.csv")
sess_real = build_sessions(df_real)
print(f"Sessions: {len(sess_real)}  |  "
      f"Attack: {sum(1 for s in sess_real.values() if s['true_label']==1)}  |  "
      f"Benign: {sum(1 for s in sess_real.values() if s['true_label']==0)}\n")

results_real = []
for rule_name, rule_set in RULES.items():
    preds = rule_predict(sess_real, rule_set)
    r = report(sess_real, preds, rule_name)
    results_real.append(r)
    print(f"{rule_name}")
    print(f"  P={r['precision']:.3f}  R={r['recall']:.3f}  F1={r['f1']:.3f}  "
          f"(TP={r['tp']} FP={r['fp']} FN={r['fn']})")
    if r["missed"]:
        print(f"  Missed: {r['missed']}")
    print()

# ── Summary table (for paper) ─────────────────────────────────────────────────
print("=" * 60)
print("PAPER TABLE — Baseline results (synthetic test set)")
print("=" * 60)
print(f"{'Method':<35} {'Precision':>10} {'Recall':>8} {'F1':>8}")
print("-" * 65)
for r in results_syn:
    print(f"{r['name']:<35} {r['precision']:>10.3f} {r['recall']:>8.3f} {r['f1']:>8.3f}")
print(f"{'GNN + Sequence model (ours)':<35} {'???':>10} {'???':>8} {'???':>8}")
print()
print("NOTE: 'Post-incident rules' represents an unfair upper bound —")
print("these rules were written AFTER reading the incident report.")
print("GuardDuty-style is the fair pre-incident comparison point.")


EVALUATION ON SYNTHETIC DATA (train distribution)
Sessions: 300  |  Attack: 100  |  Benign: 200

Minimal SIEM (3 rules)
  P=1.000  R=0.200  F1=0.333  (TP=20 FP=0 FN=80)
  Missed attack patterns: ['ListAttachedUserPolicies → GetAccountAuthorizationDetails → ListAttachedRolePolicies → UpdateAssumeRolePolicy → AssumeRole', 'GetAccountAuthorizationDetails → CreateUser → CreateAccessKey → AttachUserPolicy', 'ListAttachedRolePolicies → ListAttachedUserPolicies → UpdateAssumeRolePolicy → AssumeRole → DescribeSubnets', 'ListAttachedUserPolicies → ListAttachedRolePolicies → UpdateAssumeRolePolicy → AssumeRole → GetBucketAcl', 'AddUserToGroup → GetBucketPolicy', 'ListAttachedRolePolicies → GetAccountAuthorizationDetails → ListAttachedUserPolicies → CreateRole → AttachRolePolicy → GetBucketPolicy', 'ListAttachedUserPolicies → GetAccountAuthorizationDetails → CreateRole → AttachRolePolicy → GetSecretValue', 'CreatePolicyVersion → SetDefaultPolicyVersion → DescribeSubnets → GetBucketAcl', 'CreatePo